lib/SILOptimizer/Utils/SILInliner.cpp

```diff
@@ -225,199 +223,312 @@ class BeginApplySite {
-  for (unsigned argIdx = 0, endIdx = Args.size(); argIdx < endIdx; ++argIdx) {
-    SILValue callArg = Args[argIdx];
+
+  getBuilder().setInsertionPoint(Apply.getInstruction());
+
+  SmallVector<SILValue, 4> entryArgs;
+  entryArgs.reserve(AppliedArgs.size());
+  auto calleeConv = getCalleeFunction()->getConventions();
+  for (unsigned argIdx = 0, endIdx = AppliedArgs.size(); argIdx < endIdx;
+       ++argIdx) {
+    SILValue callArg = AppliedArgs[argIdx];
     // Insert begin/end borrow for guaranteed arguments.
     if (argIdx >= calleeConv.getSILArgIndexOfFirstParam()
         && calleeConv.getParamInfoForSILArg(argIdx).isGuaranteed()) {
-      callArg = borrowFunctionArgument(callArg, AI);
+      callArg = borrowFunctionArgument(callArg, Apply);
     }
-    auto *calleeArg = CalleeFunction->getArgument(argIdx);
-    ValueMap.insert(std::make_pair(calleeArg, callArg));
+    entryArgs.push_back(callArg);
   }
 
-  // Set up the coroutine-specific inliner if applicable.
-  auto BeginApply = BeginApplySite::get(AI, Loc.getValue(), getBuilder());
-
-  // Recursively visit callee's BB in depth-first preorder, starting with the
-  // entry block, cloning all instructions other than terminators.
-  visitSILBasicBlock(CalleeEntryBB);
-
-  // If we're inlining into a normal apply and the callee's entry
-  // block ends in a return, then we can avoid a split.
-  if (auto nonTryAI = dyn_cast<ApplyInst>(AI)) {
-    if (auto *RI = dyn_cast<ReturnInst>(CalleeEntryBB->getTerminator())) {
-      // Replace all uses of the apply instruction with the operands of the
-      // return instruction, appropriately mapped.
-      nonTryAI->replaceAllUsesWith(remapValue(RI->getOperand()));
-      return;
-    }
-  }
+  // Create the return block and set ReturnToBB for use in visitTerminator
+  // callbacks.
+  SILBasicBlock *callerBB = Apply.getParent();
+  switch (Apply.getKind()) {
+  case FullApplySiteKind::ApplyInst: {
+    auto *AI = dyn_cast<ApplyInst>(Apply);
 
-  // If we're inlining into a try_apply, we already have a return-to BB.
-  SILBasicBlock *ReturnToBB;
-
-  if (auto tryAI = dyn_cast<TryApplyInst>(AI)) {
-    ReturnToBB = tryAI->getNormalBB();
-
-  // Otherwise, split the caller's basic block to create a return-to BB.
-  } else {
-    SILBasicBlock *CallerBB = AI.getParent();
     // Split the BB and do NOT create a branch between the old and new
     // BBs; we will create the appropriate terminator manually later.
-    ReturnToBB = CallerBB->split(InsertPoint);
-    // Place the return-to BB after all the other mapped BBs.
-    if (InsertBeforeBB)
-      F.getBlocks().splice(SILFunction::iterator(InsertBeforeBB), F.getBlocks(),
-                           SILFunction::iterator(ReturnToBB));
-    else
-      F.getBlocks().splice(F.getBlocks().end(), F.getBlocks(),
-                           SILFunction::iterator(ReturnToBB));
+    ReturnToBB =
+      callerBB->split(std::next(Apply.getInstruction()->getIterator()));
```

Comment:
Is this an intentional decision to not put the return-to block in the same relative position to the inlined code?

Reply:
Okay, that makes sense.